### DISCUSSION TITLE
generateKeyPair return object with empty field

### DISCUSSION THREAD

**Discussion by giovanni9793** [panva/jose#167] [Q&A] [answered]
I don't know if I'm doing something wrong, but the function generateKeyPair does not work for me, but it seems a problem with crypto because also the following code return the same object:

```ts
import generateKeyPair from 'crypto'
import { promisify } from 'util'

const keys = promisify(generateKeyPair.generateKeyPair)
        console.log(
          JSON.stringify(
            await keys('rsa', {
              modulusLength: 2048,
              publicExponent: 0x10001,
            }),
          ),
        )
```

output:
```ts
{"publicKey":{},"privateKey":{}}
```

The code was runned on a docker machine with node 14.

**Comment by panva** [comment, marked as answer]
That's expected.

```
> JSON.stringify(crypto.generateKeyPairSync('ed25519'))
'{"publicKey":{},"privateKey":{}}'
> console.log(crypto.generateKeyPairSync('ed25519').privateKey)
PrivateKeyObject { [Symbol(kKeyType)]: 'private' }
```

See [KeyObject](https://nodejs.org/api/crypto.html#crypto_class_keyobject) docs.

**Reply by giovanni9793** [reply]
```
EDIT, same behaviour also with await
>console.log((await crypto.generateKeyPairSync('ed25519')).privateKey)
[object Object]
>console.log(JSON.stringify((await crypto.generateKeyPairSync('ed25519')).privateKey))
{}
```

Is it expected? If yes, how can I know what is the key since also "fromKeyLike" print "{}"?
```
>console.log( 'Private: ',JSON.stringify(fromKeyLike(private_key)))
Private: {}
```

**Reply by panva** [reply]
Yes it is expected. You're stringifying the return of [`fromKeyLike`](https://github.com/panva/jose/blob/main/docs/functions/_jwk_from_key_like_.fromkeylike.md) which is a [Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise), one that is resolved with a JWK object.

**Reply by giovanni9793** [reply]
Ok, thank you for the docs, so the code to use to show the private key that I have created with generateKeyPair is:
```ts
const { privateKey } = await generateKeyPair('RS256', {
          modulusLength: 2048,
        })

console.log(
  JSON.stringify(
    (privateKey as KeyObject).export({
      type: 'pkcs1',
      format: 'pem',
    }),
  ),
)
```
instead if I have a RSA key a could use crypto.createPrivateKey to obtain a KeyLike, so now that i have it I can use
```ts
id_token = await new SignJWT(baseJwt)
                  .setProtectedHeader({ alg: 'RS256' })
                  .sign(privateKey)
```
